src/core.jl

```diff
@@ -0,0 +1,167 @@
+#####################################
+### Interface for transformations ###
+#####################################
+
+# The external interface consists of transform(), Base.inv(),  compose() (or ∘),
+# transform_deriv() and transform_deriv_params()
+
+"""
+`AbstractTransformation{OutType, InType}` defines a simple interface for
+performing transformations. Subtypes should be able to apply a coordinate system
+transformation on the correct data types by overloading `transform()`, and
+usually would have the corresponding inverse transformation defined by `Base.inv()`.
+Efficient compositions can optionally be defined by `compose()` (equivalently `∘`).
+"""
+abstract AbstractTransformation{OutType, InType}
+
+# Some basic introspection
+@inline intype{OutType, InType}(::AbstractTransformation{OutType, InType}) = InType
+@inline outtype{OutType, InType}(::AbstractTransformation{OutType, InType}) = OutType
+
+# Unfortunately some operations on abstract types are difficult
+@generated function intype{T <: AbstractTransformation}(::Type{T})
+    S = T
+    while S.name != AbstractTransformation.name
+        S = super(S)
+        if S == Any
+            str = "Error determining intype of $T"
+            return :(error(str))
+        end
+    end
+
+    return :($(S.parameters[2]))
+end
+
+@generated function outtype{T <: AbstractTransformation}(::Type{T})
+    S = T
+    while S.name != AbstractTransformation.name
+        S = super(S)
+        if S == Any
+            str = "Error determining outtype of $T"
+            return :(error(str))
+        end
+    end
+
+    return :($(S.parameters[1]))
+end
+
+
+# Built-in identity transform
+immutable IdentityTransformation{T} <: AbstractTransformation{T,T}; end
+
+"""
+    transform(trans::AbstractTransformation, x)
+
+A transformation `trans` is explicitly applied to data x, returning the
+coordinates in the new coordinate system.
+"""
+function transform{OutType, InType}(trans::AbstractTransformation{OutType, InType}, x)
+    if isa(x, Vector) && eltype(x) <: InType
+        [transform(trans, point) for point in x]
```

Comment:
Two issues here:

* Shouldn't we specify `x <: Vector` for the comprehension version, rather than using `isa()`?
* Using `eltype()` seems dubious, since it'll fail for x::Vector{Any}, even if `typeof(x[i]) <: InType` for all `i`.